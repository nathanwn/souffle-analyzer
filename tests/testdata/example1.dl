/* A is a relation. */
.decl A(x: number, y: number)

/**
 * B is another relation.
**/
.decl B(x: number, y: number)

// Some rules for B.
B(x, y) :- A(x, y).
B(x, z) :- A(x, y), B(y, z).

.decl C(x: number, y: number)
C(x, y) :- A(x, y), !B(x, y).

.input A
.output C
