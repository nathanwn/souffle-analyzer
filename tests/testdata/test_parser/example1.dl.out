Document(
  range_=0:0-17:0
  BlockComment(location='example1.dl#0:0-0:22', content='/* A is a relation. */')
  RelationDeclaration(
    range_=1:0-1:29
    ResultNode(
      range_=1:6-1:7
      Identifier(location='example1.dl#1:6-1:7', val='A')
    )
    Attribute(
      range_=1:8-1:17
      ResultNode(
        range_=1:8-1:9
        Identifier(location='example1.dl#1:8-1:9', val='x')
      )
      ResultNode(
        range_=1:11-1:17
        TypeReference(
          range_=1:11-1:17
          ResultNode(
            range_=1:11-1:17
            TypeReferenceName(
              range_=1:11-1:17
              Identifier(location='example1.dl#1:11-1:17', val='number')
            )
          )
        )
      )
    )
    Attribute(
      range_=1:19-1:28
      ResultNode(
        range_=1:19-1:20
        Identifier(location='example1.dl#1:19-1:20', val='y')
      )
      ResultNode(
        range_=1:22-1:28
        TypeReference(
          range_=1:22-1:28
          ResultNode(
            range_=1:22-1:28
            TypeReferenceName(
              range_=1:22-1:28
              Identifier(location='example1.dl#1:22-1:28', val='number')
            )
          )
        )
      )
    )
  )
  BlockComment(location='example1.dl#3:0-5:3', content='/**\n * B is another relation.\n**/')
  RelationDeclaration(
    range_=6:0-6:29
    ResultNode(
      range_=6:6-6:7
      Identifier(location='example1.dl#6:6-6:7', val='B')
    )
    Attribute(
      range_=6:8-6:17
      ResultNode(
        range_=6:8-6:9
        Identifier(location='example1.dl#6:8-6:9', val='x')
      )
      ResultNode(
        range_=6:11-6:17
        TypeReference(
          range_=6:11-6:17
          ResultNode(
            range_=6:11-6:17
            TypeReferenceName(
              range_=6:11-6:17
              Identifier(location='example1.dl#6:11-6:17', val='number')
            )
          )
        )
      )
    )
    Attribute(
      range_=6:19-6:28
      ResultNode(
        range_=6:19-6:20
        Identifier(location='example1.dl#6:19-6:20', val='y')
      )
      ResultNode(
        range_=6:22-6:28
        TypeReference(
          range_=6:22-6:28
          ResultNode(
            range_=6:22-6:28
            TypeReferenceName(
              range_=6:22-6:28
              Identifier(location='example1.dl#6:22-6:28', val='number')
            )
          )
        )
      )
    )
  )
  LineComment(location='example1.dl#8:0-8:20', content=['// Some rules for B.'])
  Rule(
    range_=9:0-9:19
    RuleHead(
      range_=9:0-9:7
      ResultNode(
        range_=9:0-9:7
        RelationReference(
          range_=9:0-9:7
          ResultNode(
            range_=9:0-9:1
            RelationReferenceName(
              range_=9:0-9:1
              Identifier(location='example1.dl#9:0-9:1', val='B')
            )
          )
          Variable(location='example1.dl#9:2-9:3', ty=UnresolvedType(), name='x')
          Variable(location='example1.dl#9:5-9:6', ty=UnresolvedType(), name='y')
        )
      )
    )
    ResultNode(
      range_=9:11-9:18
      Disjunction(
        range_=9:11-9:18
        Conjunction(
          range_=9:11-9:18
          ResultNode(
            range_=9:11-9:18
            RelationReferenceClause(
              range_=9:11-9:18
              ResultNode(
                range_=9:11-9:18
                RelationReference(
                  range_=9:11-9:18
                  ResultNode(
                    range_=9:11-9:12
                    RelationReferenceName(
                      range_=9:11-9:12
                      Identifier(location='example1.dl#9:11-9:12', val='A')
                    )
                  )
                  Variable(location='example1.dl#9:13-9:14', ty=UnresolvedType(), name='x')
                  Variable(location='example1.dl#9:16-9:17', ty=UnresolvedType(), name='y')
                )
              )
            )
          )
        )
      )
    )
  )
  Rule(
    range_=10:0-10:28
    RuleHead(
      range_=10:0-10:7
      ResultNode(
        range_=10:0-10:7
        RelationReference(
          range_=10:0-10:7
          ResultNode(
            range_=10:0-10:1
            RelationReferenceName(
              range_=10:0-10:1
              Identifier(location='example1.dl#10:0-10:1', val='B')
            )
          )
          Variable(location='example1.dl#10:2-10:3', ty=UnresolvedType(), name='x')
          Variable(location='example1.dl#10:5-10:6', ty=UnresolvedType(), name='z')
        )
      )
    )
    ResultNode(
      range_=10:11-10:27
      Disjunction(
        range_=10:11-10:27
        Conjunction(
          range_=10:11-10:27
          ResultNode(
            range_=10:11-10:18
            RelationReferenceClause(
              range_=10:11-10:18
              ResultNode(
                range_=10:11-10:18
                RelationReference(
                  range_=10:11-10:18
                  ResultNode(
                    range_=10:11-10:12
                    RelationReferenceName(
                      range_=10:11-10:12
                      Identifier(location='example1.dl#10:11-10:12', val='A')
                    )
                  )
                  Variable(location='example1.dl#10:13-10:14', ty=UnresolvedType(), name='x')
                  Variable(location='example1.dl#10:16-10:17', ty=UnresolvedType(), name='y')
                )
              )
            )
          )
          ResultNode(
            range_=10:20-10:27
            RelationReferenceClause(
              range_=10:20-10:27
              ResultNode(
                range_=10:20-10:27
                RelationReference(
                  range_=10:20-10:27
                  ResultNode(
                    range_=10:20-10:21
                    RelationReferenceName(
                      range_=10:20-10:21
                      Identifier(location='example1.dl#10:20-10:21', val='B')
                    )
                  )
                  Variable(location='example1.dl#10:22-10:23', ty=UnresolvedType(), name='y')
                  Variable(location='example1.dl#10:25-10:26', ty=UnresolvedType(), name='z')
                )
              )
            )
          )
        )
      )
    )
  )
  RelationDeclaration(
    range_=12:0-12:29
    ResultNode(
      range_=12:6-12:7
      Identifier(location='example1.dl#12:6-12:7', val='C')
    )
    Attribute(
      range_=12:8-12:17
      ResultNode(
        range_=12:8-12:9
        Identifier(location='example1.dl#12:8-12:9', val='x')
      )
      ResultNode(
        range_=12:11-12:17
        TypeReference(
          range_=12:11-12:17
          ResultNode(
            range_=12:11-12:17
            TypeReferenceName(
              range_=12:11-12:17
              Identifier(location='example1.dl#12:11-12:17', val='number')
            )
          )
        )
      )
    )
    Attribute(
      range_=12:19-12:28
      ResultNode(
        range_=12:19-12:20
        Identifier(location='example1.dl#12:19-12:20', val='y')
      )
      ResultNode(
        range_=12:22-12:28
        TypeReference(
          range_=12:22-12:28
          ResultNode(
            range_=12:22-12:28
            TypeReferenceName(
              range_=12:22-12:28
              Identifier(location='example1.dl#12:22-12:28', val='number')
            )
          )
        )
      )
    )
  )
  Rule(
    range_=13:0-13:29
    RuleHead(
      range_=13:0-13:7
      ResultNode(
        range_=13:0-13:7
        RelationReference(
          range_=13:0-13:7
          ResultNode(
            range_=13:0-13:1
            RelationReferenceName(
              range_=13:0-13:1
              Identifier(location='example1.dl#13:0-13:1', val='C')
            )
          )
          Variable(location='example1.dl#13:2-13:3', ty=UnresolvedType(), name='x')
          Variable(location='example1.dl#13:5-13:6', ty=UnresolvedType(), name='y')
        )
      )
    )
    ResultNode(
      range_=13:11-13:28
      Disjunction(
        range_=13:11-13:28
        Conjunction(
          range_=13:11-13:28
          ResultNode(
            range_=13:11-13:18
            RelationReferenceClause(
              range_=13:11-13:18
              ResultNode(
                range_=13:11-13:18
                RelationReference(
                  range_=13:11-13:18
                  ResultNode(
                    range_=13:11-13:12
                    RelationReferenceName(
                      range_=13:11-13:12
                      Identifier(location='example1.dl#13:11-13:12', val='A')
                    )
                  )
                  Variable(location='example1.dl#13:13-13:14', ty=UnresolvedType(), name='x')
                  Variable(location='example1.dl#13:16-13:17', ty=UnresolvedType(), name='y')
                )
              )
            )
          )
          ResultNode(
            range_=13:20-13:28
            RelationReferenceClause(
              range_=13:20-13:28
              ResultNode(
                range_=13:21-13:28
                RelationReference(
                  range_=13:21-13:28
                  ResultNode(
                    range_=13:21-13:22
                    RelationReferenceName(
                      range_=13:21-13:22
                      Identifier(location='example1.dl#13:21-13:22', val='B')
                    )
                  )
                  Variable(location='example1.dl#13:23-13:24', ty=UnresolvedType(), name='x')
                  Variable(location='example1.dl#13:26-13:27', ty=UnresolvedType(), name='y')
                )
              )
            )
          )
        )
      )
    )
  )
  Directive(
    range_=15:0-15:8
    DirectiveQualifier(location='example1.dl#15:0-15:6', keyword='.input')
    RelationReferenceName(
      range_=15:7-15:8
      Identifier(location='example1.dl#15:7-15:8', val='A')
    )
  )
  Directive(
    range_=16:0-16:9
    DirectiveQualifier(location='example1.dl#16:0-16:7', keyword='.output')
    RelationReferenceName(
      range_=16:8-16:9
      Identifier(location='example1.dl#16:8-16:9', val='C')
    )
  )
)
